File: Danya/Lab1/main.py
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import random
import numpy as np
from pandas import DataFrame # Для красивого вывода таблицы, если есть

# ==========================================
# Настройки модели
# ==========================================
class Config:
    def __init__(self):
        self.N = 2       
        self.L = 6       
        self.Q = 3       
        self.t_mod = 40.0 
        self.dt = 0.05    
        self.server_mus = [3.5, 5.0] 
        self.initial_Z = [0.5, 1.0, 1.5]

# ==========================================
# Класс СМО
# ==========================================
class SMO:
    def __init__(self, config):
        self.cfg = config
        self.t = 0.0
        self.Z = list(self.cfg.initial_Z)
        self.U = [0.0] * self.cfg.N
        self.j = [0] * self.cfg.Q
        
        self.server_occupant = [-1] * self.cfg.N
        
        # Статистика для итогового вывода
        self.B = [0] * self.cfg.Q # Поступило
        self.P = [0] * self.cfg.Q # Отказано
        self.S = [0] * self.cfg.Q # Обслужено (успешно)
        
        # История для графиков
        self.time_history = []
        self.queue_history = [[] for _ in range(self.cfg.Q)]
        self.server_history = [[] for _ in range(self.cfg.N)]
        self.reject_events = []   
        self.arrival_events = []  

    def gen_random_interval(self):
        return random.uniform(1.5, 3.0)

    def check_task(self, k_index):
        self.B[k_index] += 1
        self.arrival_events.append((self.t, k_index))
        
        current_queue_size = sum(self.j)
        if current_queue_size < self.cfg.L:
            self.j[k_index] += 1
        else:
            displaced = False
            for low_prio in range(self.cfg.Q - 1, k_index, -1):
                if self.j[low_prio] > 0:
                    self.j[low_prio] -= 1
                    self.P[low_prio] += 1
                    self.reject_events.append((self.t, low_prio))
                    self.j[k_index] += 1
                    displaced = True
                    break
            if not displaced:
                self.P[k_index] += 1
                self.reject_events.append((self.t, k_index))

    def try_process_queue(self, server_idx):
        # Перед тем как взять новую заявку, пометим старую как обслуженную
        # (если прибор только что освободился)
        old_prio = self.server_occupant[server_idx]
        if old_prio != -1:
            self.S[old_prio] += 1

        for k in range(self.cfg.Q):
            if self.j[k] > 0:
                self.j[k] -= 1
                self.U[server_idx] = self.t + self.cfg.server_mus[server_idx]
                self.server_occupant[server_idx] = k 
                return True
        
        self.server_occupant[server_idx] = -1 
        return False

    def run(self):
        while self.t <= self.cfg.t_mod:
            self.t += self.cfg.dt
            for k in range(self.cfg.Q):
                if self.t >= self.Z[k]:
                    self.check_task(k)
                    self.Z[k] += self.gen_random_interval()
            
            for i in range(self.cfg.N):
                if self.t >= self.U[i]:
                    self.try_process_queue(i)

            self.time_history.append(self.t)
            for k in range(self.cfg.Q):
                self.queue_history[k].append(self.j[k])
            for i in range(self.cfg.N):
                self.server_history[i].append(self.server_occupant[i])

    def print_final_stats(self):
        line = "+" + "-"*11 + "+" + "-"*12 + "+" + "-"*12 + "+" + "-"*14 + "+" + "-"*14 + "+"
        print("\n" + "="*30)
        print("РЕЗУЛЬТАТЫ МОДЕЛИРОВАНИЯ")
        print("="*30)
        print(f"Время: {self.cfg.t_mod}с | Очередь: {self.cfg.L} | Приборов: {self.cfg.N}")
        print(line)
        print("| Приоритет | Поступило  |  Отказов   | Вер. отказа  | Вер. обслуж. |")
        print(line)
        
        total_B, total_P = sum(self.B), sum(self.P)
        
        for k in range(self.cfg.Q):
            p_ref = self.P[k] / self.B[k] if self.B[k] > 0 else 0
            p_srv = 1.0 - p_ref
            print(f"| {k+1:^9} | {self.B[k]:^10} | {self.P[k]:^10} | {p_ref:^12.4f} | {p_srv:^12.4f} |")
        
        print(line)
        t_p_ref = total_P / total_B if total_B > 0 else 0
        print(f"| {'ИТОГО':^9} | {total_B:^10} | {total_P:^10} | {t_p_ref:^12.4f} | {1-t_p_ref:^12.4f} |")
        print(line + "\n")

# ==========================================
# Визуализация
# ==========================================
def plot_results(model):
    t = np.array(model.time_history)
    fig, axes = plt.subplots(4, 1, figsize=(14, 10), sharex=True, 
                             gridspec_kw={'height_ratios': [1, 2, 0.6, 0.6]})
    
    colors = ['#e74c3c', '#2ecc71', '#3498db'] 
    prio_labels = ['P1 (High)', 'P2 (Mid)', 'P3 (Low)']

    # 1. Поступление
    for k in range(model.cfg.Q):
        k_times = [evt[0] for evt in model.arrival_events if evt[1] == k]
        axes[0].scatter(k_times, [k]*len(k_times), color=colors[k], marker='|', s=150, linewidth=2)
    axes[0].set_yticks(range(model.cfg.Q)), axes[0].set_yticklabels(['P1', 'P2', 'P3'])
    axes[0].set_title('Входящий поток заявок')

    # 2. Очереди
    for k in range(model.cfg.Q):
        axes[1].step(t, model.queue_history[k], where='post', color=colors[k], linewidth=2, label=prio_labels[k])
    if model.reject_events:
        rej_t, rej_p = zip(*model.reject_events)
        axes[1].scatter(rej_t, [model.cfg.L + 0.3]*len(rej_t), marker='x', color='black', label='Отказ')
    axes[1].set_ylabel('В очереди'), axes[1].set_ylim(-0.5, model.cfg.L + 1), axes[1].legend(loc='upper right')

    # 3 & 4. Приборы
    for i in range(model.cfg.N):
        ax_s = axes[2 + i]
        occ_history = np.array(model.server_history[i])
        for k in range(model.cfg.Q):
            ax_s.fill_between(t, 0, 1, where=(occ_history==k), color=colors[k], alpha=0.7, step='post')
        ax_s.set_ylabel(f'Прибор {i+1}', rotation=0, labelpad=40, va='center')

    plt.tight_layout()
    plt.show()

# ==========================================
# Старт
# ==========================================
if __name__ == "__main__":
    cfg = Config()
    model = SMO(cfg)
    model.run()
    
    # 1. Сначала выводим статистику в консоль
    model.print_final_stats()
    
    # 2. Затем показываем график
    plot_results(model)

File: Danya/Lab1/main.cpp
#include <iostream>
#include <vector>
#include <numeric>
#include <random>
#include <iomanip>

using namespace std;

// ==========================================
// Настройки модели
// ==========================================
struct Config {
    int N = 1;      // Количество приборов (маршрутов)
    int L = 6;      // Максимальная длина очереди
    int Q = 2;      // Количество источников (приоритетов)

    double t_mod = 25;     // Время моделирования
    double dt = 0.1;          // Шаг моделирования

    double mu = 1.8;          // Время обработки заявки (const)

    // Начальные Z для каждого приоритета (размер должен совпадать с Q)
    vector<double> initial_Z = {0.15, 0.625};
};

class SMO {
private:
    Config cfg;
    double t; // Текущее время (t_real)

    // Переменные состояния
    vector<double> Z; // Моменты поступления заявок (размер Q)
    vector<double> U; // Время освобождения приборов (размер N)

    // Статистика
    vector<int> B;    // Всего заявок (поступивших)
    vector<int> P;    // Количество отказов
    vector<int> j;    // Заявок в очереди по приоритетам

    // Генератор случайных чисел
    double gen_random_interval() {
        static random_device rd;
        static mt19937 gen(rd());
        static uniform_real_distribution<float> dist(1.2, 1.5);
        return dist(gen);
    }

public:
    SMO(Config config) : cfg(config) {
        t = 0.0;

        // Инициализация векторов под размер задачи
        Z = cfg.initial_Z;
        U.resize(cfg.N, 0.0); // N приборов, время 0
        B.resize(cfg.Q, 0);
        P.resize(cfg.Q, 0);
        j.resize(cfg.Q, 0);
    }

    // Логика постановки в очередь (check_task)
    void check_task(int k_index) {
        int current_queue_size = accumulate(j.begin(), j.end(), 0);

        // Если есть место в очереди
        if (current_queue_size < cfg.L) {
            j[k_index]++;
        }
        // Если очередь полна, пытаемся вытеснить менее важную заявку
        else {
            bool displaced = false;
            // Ищем заявку с более низким приоритетом (индекс больше), чем k_index
            // Идем с конца (самый низкий приоритет)
            for (int low_prio = cfg.Q - 1; low_prio > k_index; --low_prio) {
                if (j[low_prio] > 0) {
                    j[low_prio]--;      // Выкидываем низкоприоритетную
                    P[low_prio]++;      // Записываем ей отказ
                    j[k_index]++;       // Ставим новую
                    displaced = true;
                    break;
                }
            }

            if (!displaced) {
                // Если вытеснить некого (все важнее или такие же), отказываем текущей
                P[k_index]++;
            }
        }
    }

    // Логика извлечения из очереди (trow_task) и загрузки прибора
    // Возвращает true, если прибор взял задачу
    bool try_process_queue(int server_idx) {
        // Проверяем очередь, начиная с высшего приоритета (0)
        for (int k = 0; k < cfg.Q; ++k) {
            if (j[k] > 0) {
                j[k]--; // Убираем из очереди
                U[server_idx] = t + cfg.mu; // Занимаем прибор
                return true;
            }
        }
        return false;
    }

    void run() {
        while (t <= cfg.t_mod) {
            t += cfg.dt;

            // 1. Проверка поступления новых заявок (цикл по всем источникам)
            for (int k = 0; k < cfg.Q; ++k) {
                if (t >= Z[k]) {
                    B[k]++;             // Увеличиваем счетчик поступивших
                    check_task(k);      // Пробуем поставить в очередь
                    Z[k] += gen_random_interval(); // Планируем следующую
                }
            }

            // 2. Проверка освобождения приборов (цикл по всем приборам)
            for (int i = 0; i < cfg.N; ++i) {
                // Если прибор свободен (время освобождения <= текущему)
                // ИЛИ только что освободился
                if (t >= U[i]) {
                    // Пытаемся взять задачу из очереди
                    try_process_queue(i);
                }
            }
        }
    }

    void print_stats() {
        cout << fixed << setprecision(4);
        cout << "=== Результаты моделирования ===" << endl;
        cout << "Приборов (N): " << cfg.N << ", Очередь (L): " << cfg.L << endl;

        for (int k = 0; k < cfg.Q; ++k) {
            double p_refusal = (B[k] > 0) ? (double)P[k] / B[k] : 0.0;

            cout << "----------------------------" << endl;
            cout << "Приоритет " << (k + 1) << ":" << endl;
            cout << "  Поступило (B[" << k+1 << "]): " << B[k] << endl;
            cout << "  Отказов   (P[" << k+1 << "]): " << P[k] << endl;
            cout << "  Вероятность отказа:  " << p_refusal << endl;
            cout << "  Вероятность обслуж.: " << (1.0 - p_refusal) << endl;
        }
        cout << "----------------------------" << endl;
    }
};

int main() {
    setlocale(LC_ALL, "Russian");

    // Настройка параметров перед запуском
    Config myConf;

    SMO model(myConf);
    model.run();
    model.print_stats();

    return 0;
}
